torch/csrc/api/src/serialize/output-archive.cpp

```diff
@@ -19,7 +19,11 @@ OutputArchive::OutputArchive(std::shared_ptr<jit::script::CompilationUnit> cu)
       module_("__torch__.Module", cu_, /*shouldMangle=*/true) {}
 
 void OutputArchive::write(const std::string& key, const c10::IValue& ivalue) {
-  module_.register_attribute(key, ivalue.type(), ivalue);
+  if(!(ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton())) {
```

Comment:
For this if-branch, we are checking whether the tensor _is_ undefined, and for that we probably should do `ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton()` instead

Reply:
It would also be awesome add this to our test suite:
```cpp
{
  auto tempfile = c10::make_tempfile();

  auto ivalue = torch::IValue(torch::Tensor());

  ASSERT_TRUE(ivalue.isTensor());
  ASSERT_FALSE(ivalue.toTensor().defined());
  
  torch::serialize::OutputArchive output_archive;
  output_archive.write("value", ivalue);
  output_archive.save_to(tempfile.name);

  torch::serialize::InputArchive input_archive;
  input_archive.load_from(tempfile.name);
  c10::IValue ivalue_out;
  input_archive.read("value", ivalue_out);
  
  ASSERT_TRUE(ivalue_out.isTensor());
  ASSERT_FALSE(ivalue_out.toTensor().defined());
}
```
Which can guarantee the correctness of undefined tensor serialization/deserialization :D

Reply:
Don't forget about changing this to
```suggestion
  if (ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton()) {
```
otherwise it's doing the opposite thing :)